redhat-cop/pelorus
# Better project matching logic

## Describe the behavior changes introduced in this PR
There are three fixes here.  The first fixes the parsing of repository projects by selecting the last token for the project name.  
Second, is an addition of type Docker for code builds.
Last, when a search by project name is made to the Gitlab API, this change will attempt to find a project in the search results that matches the Git URL for the repository.

A bit more details on the changes here since Eric and I just had a conversation about this.

First, the original code would call to the Gitlab API with a project name and select the first match in the returned list.  This has been enhanced to take the repo_url from the CommitMetric and attempt to match the project which has an equivalent ssh_url_to_repo or http_url_to_repo.  See the [Gitlab API](https://docs.gitlab.com/ee/api/projects.html#get-single-project) for what is returned in the project object.  The only other possibility on the Project object returned by the API is the web_url property which I could add but have not in this commit.  Note that if we go through all of the projects and there are no equivalency matches, then the project will be set to None.  The Gitlab API doesn't document the Project object and all of its attributes; however, it seems that this equivalency test should be sufficient as forked projects would have different urls from the parent project.  Also, the search method only returns 20 results by default, so I added recursion to walk through the paged results.

Secondly, regarding partial matches.  I'm curious to your thoughts here as to how this would happen?  Would it be that the BuildConfig is created with only part of the url in spec.source.git.uri?  


## Linked Issues?

resolves #208 and #212 and #215 

## Testing Instructions

Use a project that is a grandchild of the group like https://gitlab.com/BhfLab/osd/leia.git and make sure that the correct project name is set for the CommitTime metric.

Make sure that when the Gitlab api search results come back that the correct project is matched from the search results.

@redhat-cop/mdt
